File: java8/preprocessor/preprocessor.c
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <ctype.h>

#define BUFFER_START 100
#define REFERENCE_SEPARATOR ":REF:"

static void *xrealloc(void *ptr, size_t size)
{
	ptr = realloc(ptr, size);
	if(!ptr) {
		fprintf(stderr, "Out of memory\n");
		exit(1);
	}
	return ptr;
}

static void *xmalloc(size_t size)
{
	void *ptr = malloc(size);
	if(!ptr) {
		fprintf(stderr, "Out of memory\n");
		exit(1);
	}
	return ptr;
}

static size_t locate_last_nonspace(char *text, size_t start_pos)
{
	size_t pos;
	if(start_pos == 0)
		return start_pos;
	pos = start_pos - 1;
	while(isspace(text[pos]) && pos != 0) 
		pos--;
	
	return pos;
}

static void locate_array_start(char *text, size_t *current_pos)
{
	if(*current_pos == 0)
		return;

	*current_pos = *current_pos - 1;
	while(text[*current_pos] != '[' && *current_pos != 0) 
		*current_pos = *current_pos - 1;

	return;
		
	
}

static void locate_generic_start(char *text, size_t *current_pos) 
{
	if(*current_pos == 0)
		return;
	while(1) {
		*current_pos = *current_pos - 1;
		if(*current_pos == 0) 
			return;
		else if(text[*current_pos] == '<')
			return;
		else if(text[*current_pos] == '>') 
			locate_generic_start(text, current_pos);
		
	}
}

static void locate_generic_method_references(char *text, size_t **insert_indexes)
{	
	size_t current_size = BUFFER_START;
	size_t i = 0;
	size_t current_pos;
	size_t buffer_pos = 0;
	*insert_indexes = xmalloc(sizeof(size_t) * current_size);
	(*insert_indexes)[0] = 0;
	while(text[i + 2]) {
		if(text[i] == ':' && text[i+1] == ':' && text[i+2] != ':') {
			current_pos = locate_last_nonspace(text, i);
			while(text[current_pos] == ']') {
				locate_array_start(text, &current_pos);
				if(current_pos == 0)
					return;
				current_pos = locate_last_nonspace(text, current_pos);
			}

			if(text[current_pos] == '>') {
				locate_generic_start(text, &current_pos);
				if(current_pos == 0)
					return;
				else if(text[current_pos] == '<') {
					if(buffer_pos + 1 == current_size) {
						current_size *= 2;
						*insert_indexes = xrealloc(*insert_indexes, sizeof(size_t) * current_size);
					}
					(*insert_indexes)[buffer_pos] =  current_pos;
					buffer_pos++;
					(*insert_indexes)[buffer_pos] = 0;
				}
				
			}
			else {
				i++;
			}
		}
		i++;
	}
}

int main(int argc, char *argv[])
{
	FILE *f;
	size_t current_size;
	char *text;
	size_t index;
	int character;
	size_t i;
	size_t ref_index = 0;
	size_t *insert_indexes;

	if(argc != 3) {
		fprintf(stderr, "Error, incorrect input arguments\n");
		exit(1);
	}

	f = fopen(argv[1], "r");
	if(!f) {
		fprintf(stderr, "Error, incorrect input file path\n");
		exit(1);
	}

	current_size = BUFFER_START;
	text = xmalloc(current_size);
	index = 0;

	while((character = fgetc(f)) != EOF) {
		text[index] = character;
		index++;
		if(index == current_size) {
			current_size *= 2;
			text = xrealloc(text, current_size);
		}
	}

	fclose(f);
	text[index] = '\0';
	if(index < 2) 
		return 0;

	f = fopen(argv[2], "w");

	locate_generic_method_references(text, &insert_indexes);

	for(i = 0; i < index; i++) {
		if(i == insert_indexes[ref_index] && insert_indexes[ref_index] != 0) {
			fprintf(f, "%s", REFERENCE_SEPARATOR);
			ref_index++;
		}
		fprintf(f, "%c", text[i]);
	}

	fclose(f);

	free(insert_indexes);
	free(text);
	return 0;
}

File: java8/preprocessor/Preprocessor.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.util.ArrayList;


public class Preprocessor {

	public static final String REFERENCE_SEPARATOR = ":REF:";
	private static int current_pos;
	
	public static void main(String[] args) {
		StringBuilder text = new StringBuilder();
		int character = 0;
		int i;
		int ref_index = 0;
		ArrayList<Integer> insert_indexes = new ArrayList<Integer>();
		Reader f = null;
		BufferedWriter writer = null;
		
		if(args.length != 2) {
			System.out.println("Error, incorrect input arguments");
			System.exit(1);
		}
		try {
			f = new BufferedReader(new FileReader(args[0]));
		} catch(Exception e) {
			System.out.println("Error, wrong input file path");
			System.exit(1);
		}



		try {
			character = f.read();
			while(character != -1) {
				text.append((char)character);
				character = f.read();
			}
			f.close();
		} catch (IOException e1) {
			// TODO Auto-generated catch block
			e1.printStackTrace();
		}

		if(text.length() < 2) 
			return;


		try {
		    writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(args[1]), "utf-8"));
		    
		} catch (IOException ex) {
			System.out.println("Error, wrong output file path");
			System.exit(1);
		} 

		locate_generic_method_references(text, insert_indexes);

		for(i = 0; i < text.length(); i++) {
			if(ref_index != insert_indexes.size() && i == insert_indexes.get(ref_index)) {
				try {
					writer.write(REFERENCE_SEPARATOR);
				} catch (IOException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}
				ref_index++;
			}
			try {
				writer.write(text.charAt(i));
			} catch (IOException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}

		try {
			writer.close();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
		return;
	}

	private static int locate_last_nonspace(StringBuilder text, int start_pos)
	{
		int pos;
		if(start_pos == 0) {
			return start_pos;
		}
		pos = start_pos - 1;
		while(Character.isWhitespace(text.charAt(pos)) && pos != 0) {
			pos--;
		}
		
		return pos;
	}

	static void locate_array_start(StringBuilder text) {
		if(current_pos == 0) {
			return;
		}

		current_pos = current_pos - 1;
		while(text.charAt(current_pos) != '[' && current_pos != 0) {
			current_pos = current_pos - 1;
		}

		return;	
	}

	static void locate_generic_start(StringBuilder text) 
	{
		if(current_pos == 0)
			return;
		while(true) {
			current_pos = current_pos - 1;
			if(current_pos == 0) 
				return;
			else if(text.charAt(current_pos) == '<')
				return;
			else if(text.charAt(current_pos) == '>') 
				locate_generic_start(text);
			
		}
	}

	static void locate_generic_method_references(StringBuilder text, ArrayList<Integer> insert_indexes)
	{	
		for(int i = 0; i < text.length() - 2; i++) {
			if(text.charAt(i) == ':' && text.charAt(i+1) == ':' && text.charAt(i+2) != ':') {
				current_pos = locate_last_nonspace(text, i);
				while(text.charAt(current_pos) == ']') {
					locate_array_start(text);
					if(current_pos == 0)
						return;
					current_pos = locate_last_nonspace(text, current_pos);
				}

				if(text.charAt(current_pos) == '>') {
					locate_generic_start(text);
					if(current_pos == 0)
						return;
					else if(text.charAt(current_pos) == '<') {
						insert_indexes.add(current_pos);
					}
					
				}

			}

		}
	}

}